src/procedures/proc_shortest_path.c

```diff
@@ -0,0 +1,291 @@
+	// Reverse path such that source node is at the beginning.
+	Path_Reverse(path);
+	SIArray_Reverse(cost);
+
+	*path_output = SI_Path(path);
+	*cost_output = cost;
+}
+
+ProcedureResult Proc_ShortestPathInvoke(ProcedureCtx *ctx, const SIValue *args) {
+	if(array_len((SIValue *)args) != 5) return PROCEDURE_ERR;
+
+	ctx->privateData = NULL;
+
+	Node *startNode = (Node *)(args[0].ptrval);
+	Node *endNode = (Node *)(args[1].ptrval);
+	const char *weightProperty = args[2].stringval;
+	const char *relationship = args[3].stringval;
+	double defaultValue = args[4].doubleval;
+
+	GrB_Matrix M = _BuildWeightMatrix(relationship, weightProperty, defaultValue);
+	if(M == GrB_NULL) return PROCEDURE_ERR;
+
+	ShortestPath_Ctx *pdata = rm_malloc(sizeof(ShortestPath_Ctx));
+	pdata->M = M;
+	pdata->endNode = endNode;
+	pdata->startNode = startNode;
+	pdata->relationship = relationship;
+	pdata->defaultValue = defaultValue;
+	pdata->weightProperty = weightProperty;
+	pdata->pathProduced = false;
+
+	pdata->output = array_new(SIValue, 4);
+	pdata->output = array_append(pdata->output, SI_ConstStringVal("path"));
+	pdata->output = array_append(pdata->output, SI_NullVal()); // Place holder.
+	pdata->output = array_append(pdata->output, SI_ConstStringVal("cost"));
+	pdata->output = array_append(pdata->output, SI_NullVal()); // Place holder.
+
+	ctx->privateData = pdata;
+	return PROCEDURE_OK;
+}
+
+SIValue *Proc_ShortestPathStep(ProcedureCtx *ctx) {
+	assert(ctx->privateData);
+
+	SIValue *res = NULL;
+	ShortestPath_Ctx *pdata = (ShortestPath_Ctx *)ctx->privateData;
+	Node *s = pdata->startNode;
+	Node *t = pdata->endNode;
+	NodeID s_id = ENTITY_GET_ID(s);
+	NodeID t_id = ENTITY_GET_ID(t);
+	GrB_Vector d = GrB_NULL;    // Pointer to the vector of distance
+	GrB_Vector pi = GrB_NULL;   // Pointer to the vector of parent
+	GrB_Vector h = GrB_NULL;    // Pointer to the vector of hops
+
+	if(pdata->pathProduced) goto cleanup;
+	// Mark this call to Step, such that additional calls will return NULL.
+	pdata->pathProduced = true;
+
+	GrB_Info info = LAGraph_BF_full(&d, &pi, &h, pdata->M, s_id);
+
+	if(info != GrB_SUCCESS) {
+		/* Failed to run Bellman-Ford single source shortest paths.
+		 * return NULL. */
+		goto cleanup;
+	}
+
+	// Determine number of hops on shortest path.
+	uint64_t hops;
+	GrB_Vector_extractElement_UINT64(&hops, h, t_id);
+	if(hops == 0) {
+		// No path from s to t.
+		goto cleanup;
+	}
+
+	// Set result.
+	res = pdata->output;
+
+	// s and t are connected, construct PATH object.
+	SIValue path;
+	SIValue cost;
+	_ConstructPath(&path, &cost, pi, h, d, pdata);
+	pdata->output[1] = path;
+	pdata->output[3] = cost;
+
+cleanup:
+	if(d) GrB_free(&d);
+	if(pi) GrB_free(&pi);
+	if(h) GrB_free(&h);
+
+	// Mark ourselves as depleted.
+	return res;
+}
+
+ProcedureResult Proc_ShortestPathFree(ProcedureCtx *ctx) {
+	// Clean up.
+	if(ctx->privateData) {
+		ShortestPath_Ctx *pdata = ctx->privateData;
+		GrB_free(&pdata->M);
+		array_free(pdata->output);
```

Comment:
free the path and the cost array?

Reply:
freeing the `path` will cause a double free, as we're passing the path to our caller.

Reply:
who is the owner of the path and the cost array?
just asking for void a situation of leak/double free

Reply:
`proc_shortest_path` is obviously the producer of these objects, but once it passes them to its caller it loses its ownership, and the responsible of freeing them is passed on to that caller.
At least this is how I see it.